src/uTensor/ops/Transpose.hpp

```diff
@@ -0,0 +1,123 @@
+    TensorStrides input_strides = TensorStrides(input_shape);
+
+    Tensor& output_tensor = outputs[output].tensor();
+
+    // Create a placeholder to calculate the output shape
+    // Normally this would reference output shape, but since this could (usually would) be referencing the input, let's keep a dedicated value
+    TensorShape output_shape = TensorShape(1,1,1,1);
+    TensorStrides output_strides = TensorStrides(output_shape);
+    TensorShape offsets = TensorShape(input_shape.num_dims());
+
+    for (size_t i = 0; i < 4; ++i) { 
+      output_shape[i] = 0;
+      output_strides[i] = 0;
+
+      // Offsets are used to avoid multiple for loops
+      offsets[i] = 0;
+    }
+
+    for (size_t i = 0; i < (size_t) input_shape.num_dims(); ++i) { 
+      output_shape[_axes[i]] = input_shape[i];
+
+      // output_strides(i) is derived from axes and input_strides
+      output_strides[_axes[i]] = input_strides[i];
+    }
+    
+    // Output shape can be asserted once the transform 
+    // effect has been determined
+    output_shape.update_dims();
+    output_tensor->resize(output_shape);
+
+    // Perform some basic checks
+    if (input_tensor->num_elems() != output_tensor->num_elems()){
+        uTensor_printf("inconsistent input and output shape for reshape\n");
+        Context::get_default_context()->throwError(new InvalidReshapeError);
+        return;
+    }    
+    if (input_tensor->get_type() != output_tensor->get_type()){
+        uTensor_printf("inconsistent input and output data type for reshape\n");
+        Context::get_default_context()->throwError(new InvalidTensorDataTypeError);
+        return;
+    }
+    if (!_check_input_shape()){
+        Context::get_default_context()->throwError(new InvalidTensorDataTypeError);
+        return;
+    }
+
+    // copy data
+    for (uint32_t i = 0; i < input_tensor->num_elems(); ++i) { 
+        // Index of the source value, must be calculated
+        // using the output strides and output shape
+        uint32_t idx = 0;
+        for (uint32_t j = 0; j < output_shape.num_dims(); j++) {
+            idx += offsets[j] * output_strides[j];
+        }
+
+        // this is not copy: `output_tensor(i) = input_tensor(i);`
+        output_tensor(i) = static_cast<Tin>(input_tensor(idx));
+
+        // Update offsets, to iterate sequentially along strides
+        // in the order of axes
+        for (int32_t j = output_shape.num_dims() - 1; j >= 0; j--) {
+            offsets[j] = (offsets[j] + 1) % (output_shape[j]);
+            if( offsets[j] > 0 ) {
+                break;
+            }
+        }        
+    }  
+
+  }
+private:
+  TensorShape _axes;
+
+  bool _check_input_shape(){
+    const Tensor& input_tensor = inputs[input].tensor();
+    const TensorShape& shape = input_tensor->get_shape();
+    uint8_t num_dims = shape.num_dims();
+    for (int i = 0; i < num_dims; ++i){
+      if (shape[i] < 0) {
+        uTensor_printf("the output shape must be all positive\n");
```

Comment:
Throw a utensor error here? @dboyliao we can add this as a todo

Reply:
I think @victorromeo has done it.

However, I find the exact same function is used in `ReshapeOp`.
Maybe we can extract it out and make it a general util function across `uTensor`.

Reply:
But I'm ok to leave it as is for now.

Reply:
Let’s go ahead and merge it as is then and I can migrate this into utilities tomorrow. Nice thing about having a newborn child is I get a lot of free time to hack :D

Reply:
I know.
Make sure your child and wife are of top priority.
That's much more important than this project. :)